hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/basic/AggregationResultTest.java

```diff
@@ -0,0 +1,155 @@
+		}
+)
+public class AggregationResultTest {
+	@BeforeEach
+	public void setUp(EntityManagerFactoryScope scope) {
+		scope.inTransaction( entityManager -> {
+			Product product = new Product();
+			product.setId( "product1" );
+			product.setPrice( 1.23d );
+			product.setQuantity( 1000 );
+			product.setPartNumber( ( (long) Integer.MAX_VALUE ) + 1 );
+			product.setRating( 1.999f );
+			product.setSomeBigInteger( BigInteger.valueOf( 987654321 ) );
+			product.setSomeBigDecimal( BigDecimal.valueOf( 987654.32 ) );
+			entityManager.persist( product );
+		} );
+	}
+
+	@AfterEach
+	public void tearDown(EntityManagerFactoryScope scope) {
+		scope.inTransaction( entityManager -> {
+			entityManager.createQuery( "delete Product" ).executeUpdate();
+		} );
+	}
+
+	/**
+	 * Sum of Longs should return a Long
+	 */
+	@Test
+	public void testSumOfLongs(EntityManagerFactoryScope scope) {
+		scope.inTransaction( entityManager -> {
+			CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+			CriteriaQuery<Long> criteria = builder.createQuery( Long.class );
+			Root<Product> productRoot = criteria.from( Product.class );
+			criteria.select( builder.sum( productRoot.get( Product_.partNumber ) ) );
+			Object sumResult = entityManager.createQuery( criteria ).getSingleResult();
+			assertReturnType( Long.class, sumResult );
+		} );
+	}
+
+	/**
+	 * Sum of Integers should return a Long
+	 */
+	@Test
+	public void testSumOfIntegers(EntityManagerFactoryScope scope) {
+		scope.inTransaction( entityManager -> {
+			CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+			CriteriaQuery<Long> criteria = builder.createQuery( Long.class );
+			Root<Product> productRoot = criteria.from( Product.class );
+			criteria.select( builder.sumAsLong( productRoot.get( Product_.quantity ) ) );
+			Object sumResult = entityManager.createQuery( criteria ).getSingleResult();
+			assertReturnType( Long.class, sumResult );
+		} );
+	}
+
+	/**
+	 * Sum of Doubles should return a Double
+	 */
+	@Test
+	public void testSumOfDoubles(EntityManagerFactoryScope scope) {
+		scope.inTransaction( entityManager -> {
+			CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+			CriteriaQuery<Double> criteria = builder.createQuery( Double.class );
+			Root<Product> productRoot = criteria.from( Product.class );
+			criteria.select( builder.sum( productRoot.get( Product_.price ) ) );
+			Object sumResult = entityManager.createQuery( criteria ).getSingleResult();
+			assertReturnType( Double.class, sumResult );
+		} );
+	}
+
+	/**
+	 * Sum of Floats should return a Double
+	 */
+	@Test
+	public void testSumOfFloats(EntityManagerFactoryScope scope) {
+		scope.inTransaction( entityManager -> {
+			CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+			CriteriaQuery<Double> criteria = builder.createQuery( Double.class );
+			Root<Product> productRoot = criteria.from( Product.class );
+			criteria.select( builder.sumAsDouble( productRoot.get( Product_.rating ) ) );
```

Comment:
It would fail and that is why I asked the question on Zulip. I think we broke the backward compatibility here. As per the JPA interface of javax.persistence.criteria.CriteriaBuilder, it implies we should allow this:
<img width="577" alt="Screen Shot 2021-06-01 at 12 29 25" src="https://user-images.githubusercontent.com/8211458/120359021-38b2da00-c2d5-11eb-9490-f83f0451efec.png">

End user of Criteria would feel so confused for they should rely on the API interface code artifact as opposed to JPA Spec doc. Again, @sebersole what is your input here?

Reply:
I revert back the two broken testing cases into org.hibernate.jpa.test.criteria.basic.AggregationResultTest. The float summing test case will trigger the following excpetion:

Caused by: org.hibernate.query.QueryTypeMismatchException: Specified result type [java.lang.Float] did not match Query selection type [java.lang.Double]